src/psbt/finalizer.rs

```diff
@@ -347,29 +359,46 @@ pub fn finalize_helper<C: secp256k1::Verification>(
 
     // Actually construct the witnesses
     for index in 0..psbt.inputs.len() {
-        let (witness, script_sig) = {
-            let spk = get_scriptpubkey(psbt, index).map_err(|e| Error::InputError(e, index))?;
-            let sat = PsbtInputSatisfier::new(&psbt, index);
+        finalize_input(psbt, index, secp, allow_mall)?;
+    }
+    // Double check everything with the interpreter
+    // This only checks whether the script will be executed
+    // correctly by the bitcoin interpreter under the current
+    // psbt context.
+    interpreter_check(&psbt, secp)?;
+    Ok(())
+}
 
-            if spk.is_v1_p2tr() {
-                // Deal with tr case separately, unfortunately we cannot infer the full descriptor for Tr
-                let wit = construct_tap_witness(spk, &sat, allow_mall)
-                    .map_err(|e| Error::InputError(e, index))?;
-                (wit, Script::new())
-            } else {
-                // Get a descriptor for this input.
-                let desc = get_descriptor(&psbt, index).map_err(|e| Error::InputError(e, index))?;
+pub(super) fn finalize_input<C: secp256k1::Verification>(
+    psbt: &mut Psbt,
+    index: usize,
+    secp: &Secp256k1<C>,
+    allow_mall: bool,
+) -> Result<(), super::Error> {
+    let (witness, script_sig) = {
+        let spk = get_scriptpubkey(psbt, index).map_err(|e| Error::InputError(e, index))?;
+        let sat = PsbtInputSatisfier::new(&psbt, index);
 
-                //generate the satisfaction witness and scriptsig
-                if !allow_mall {
-                    desc.get_satisfaction(PsbtInputSatisfier::new(&psbt, index))
-                } else {
-                    desc.get_satisfaction_mall(PsbtInputSatisfier::new(&psbt, index))
-                }
-                .map_err(|e| Error::InputError(InputError::MiniscriptError(e), index))?
+        if spk.is_v1_p2tr() {
+            // Deal with tr case separately, unfortunately we cannot infer the full descriptor for Tr
+            let wit = construct_tap_witness(spk, &sat, allow_mall)
+                .map_err(|e| Error::InputError(e, index))?;
+            (wit, Script::new())
+        } else {
+            // Get a descriptor for this input.
+            let desc = get_descriptor(&psbt, index).map_err(|e| Error::InputError(e, index))?;
+
+            //generate the satisfaction witness and scriptsig
+            if !allow_mall {
+                desc.get_satisfaction(PsbtInputSatisfier::new(&psbt, index))
+            } else {
+                desc.get_satisfaction_mall(PsbtInputSatisfier::new(&psbt, index))
```

Comment:
In 815d012f2280e89d609c57c2d7357bd86b69c545:

Can't these use `sat` as well rather than creating new input satisfiers?

Reply:
seems like it should work, no borrowchecker issues i can think of.